### ISSUE TITLE
Question about conjugate gradient

### ISSUE THREAD

**Issue by thomas-yu-epfl** [utcsilab/deepinpy#22]
Hi, thanks very much for this project, it was very useful for getting me started in MRI image reconstruction/machine learning. 

I had a question about the implementation of the conjugate gradient implementation:
specifically, you remove the variable r from the computational graph; for what reason do you do this?

I ask because I am trying to train an unrolled network (similar to MODL), but I am finding that my conj grad block adds enormously to the GPU memory during training (without :2GB, with:42GB). However, when I tried to replicate this removal in my code, I get an error that this can only be done to leaf variables. I'm not so familiar in pytorch so I am a bit lost how to handle the memory issue.

I would appreciate any light you could shed on the situation.

Thanks!

**Comment by jtamir** [comment]
Hi --

To be honest I cannot remember. My old commit message said it was to "fix some conjgrad memory leaks". I recall having an error related to leaf nodes, but I am not able to reproduce.

Based on the current code, we realize that explicitly removing "r" from the graph is not necessary since it will get reassigned immediately after. Furthermore, lambda can be negative since the unrolled optimization is different from true CG. We will change this in an update soon.

I have not noticed any issues with memory. Are you running many unrolls of CG? The implementation of MoDL in the repo should work correctly.

**Comment by thomas-yu-epfl** [comment]
Thanks for your reply:

I am trying to replicate a paper in pytorch where they have a network similar to MoDL, and they do an unrolled network with 10 iterations. Their implementation is in tensorflow/they say they have done their training on a GPU with 10Gb of memory, so this is why I thought there might be a problem with my implementation, since 10 iterations of CG uses much more memory in my implementation.

**Comment by jtamir** [comment]
It may be possible to reduce the memory usage of CG by a small amount, but I do not think there are extra variables on the computational graph that aren't necessary. It is hard to say if your observed behavior is normal without knowing the details of your problem size or setup.  

Here is what I see with the DeepInPy implementation of MoDL:
image size: [256, 232]
6 coils
batch size of 4
ResNet Model of 4M parameters
4 unrolls
10 CG iterations

GPU memory usage is 13154MiB

**Comment by thomas-yu-epfl** [comment]
Thanks for the info! I do notice that the memory usage scales with the number of unrolls; so I think my memory usage makes sense since you use less than half the unrolls that I used but with 13GB of memory used. Also my data has 64 coils so this also probably has a significant impact. So I guess the answer is to use less unrolls and perhaps less precision on the data?

Thank you very much for the help! I'll close this issue as solved.

**Comment by jtamir** [comment]
If you are using MoDL then the number of coils should be irrelevant. I recommend only passing the adjoint to your forward pass. This may not be the way I have it implemented.

Jon
www.jtsense.com
________________________________

**Comment by thomas-yu-epfl** [comment]
Ok, thanks for the tip!
